sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/UploadHandler.java

```diff
@@ -0,0 +1,337 @@
+            p.second.cancel();
+        }
+        this.uploadHandlerListener.onError(failedPair.first.getStagingError());
+        this.getLooper().quit();
+    }
+
+    /**
+     * Handles the blocks commit completion message received by the looper.
+     *
+     * This stage notifies the completion of file upload to {@link UploadHandlerListener}
+     * and terminates the handler.
+     */
+    private void handleCommitCompleted() {
+        this.uploadHandlerListener.onUploadProgress(this.blob.fileSize, this.blob.fileSize);
+        this.uploadHandlerListener.onComplete();
+        this.getLooper().quit();
+    }
+
+    /**
+     * Handles the blocks commit failed message received by the looper.
+     *
+     * This stage notifies the failure to {@link UploadHandlerListener} and terminates
+     * the handler.
+     */
+    private void handleCommitFailed() {
+        this.uploadHandlerListener.onError(this.blob.getCommitError());
+        this.getLooper().quit();
+    }
+
+    /**
+     * Check whether stop token is signalled, if so park the work and quit the looper.
+     */
+    private void finalizeIfStopped() {
+        if (this.uploadStopToken.isStopped()) {
+            Log.v(TAG, "finalizeIfStopped(): Stop request received, finalizing");
+            for (Pair<BlockUploadEntity, ServiceCall> p : this.runningBlockUploads.values()) {
+                p.second.cancel();
+            }
+            UploadInterruptState interruptState = this.db.uploadDao().getUploadInterruptState(this.uploadId);
+            Log.v(TAG, "finalizeIfStopped: Stop request reason (NONE == Stop requested by SYSTEM): " + interruptState);
+            switch (interruptState) {
+                case NONE:
+                    this.uploadHandlerListener.onSystemPaused();
+                    break;
+                case USER_PAUSED:
+                    this.uploadHandlerListener.onUserPaused();
+                    break;
+                case USER_CANCELLED:
+                    this.db.uploadDao().updateUploadInterruptState(this.uploadId, UploadInterruptState.PURGE);
+                    this.uploadHandlerListener.onError(new UploadCancelledException(this.uploadId));
+            }
+            this.getLooper().quit();
+        }
+    }
+
+    /**
+     * Starts the block upload async operations.
+     *
+     * @param blocks the blocks to be staged (uploaded).
+     */
+    private void stageBlocks(List<BlockUploadEntity> blocks) {
+        for (BlockUploadEntity block : blocks) {
+            this.finalizeIfStopped();
+            Log.v(TAG, "stageBlocks(): Uploading block:" + block.blockId + threadName());
+            ServiceCall call = this.blobClient.stageBlock(this.blob.containerName,
+                this.blob.blobName,
+                block.blockId,
+                block.getBlockContent(),
+                null, new com.azure.android.core.http.Callback<Void>() {
```

Comment:
Nit: Can we import this type?

Reply:
sure we can. Will update the PR.

Reply:
It seems we need this to be fully qualified. The reason is, the parent `Handler` class has a [Callback](https://developer.android.com/reference/android/os/Handler.Callback) contract in it, which take more precedence.